exquery-restxq-specification/restxq-2.0-specification.html

```diff
@@ -402,6 +402,10 @@ <h5>Method Annotation</h5>
                 interpretation of the request body is similar to that of the
                 <a href="http://expath.org/spec/http-client">EXPath HTTP Client</a>:
                 <ol>
+                  <li>Multiple function parameters MUST only be used with a <code>multipart/*</code> Media Type. This media type MAY be implicitly set, or set explicitly using a <code>%rest:consumes</code> annotation. Implementations MUST throw an error if more than one parameter is mapped where any other Media Type is defined using <code>%rest:consumes</code>.</li>
+                  <li>For the <code>multipart/alternative</code> Media type, the order of request body parts SHOULD be processed in reverse in order to preserve the last bodypart having the highest priority.  For all other <code>multipart/*</code> requests, body part order SHOULD be preserved.</li>
+                  <li>Where multiple function parameters are defined, parameters are injected from the body parts of the multipart request until the final string literal is defined.  Each parameter must individually meet this same list of constraints.</li>
+                  <li>The remaining body parts are injected to the function parameter corresponding to the final string literal as an array.  Implementations MAY cast this array to a sequence if it matches cardinality with the function parameter.</li>
```

Comment:
I think the parts about implicit/explicit media types seems sensible.

However, I want to check if I understand (correctly or not) your proposal about how multipart requests are decomposed to function parameters... I think some examples would be very helpful not only for me, but also for the consumers or the spec.



1. So do I understand correctly, you can write this:
    ```xquery
    declare
       %reset:POST("{$body}")
    function my:example1($body as item()*) {
    ...
    }
    ```

    1. If this is not a multipart request, presumably the type of `$body` stays as it was in RESTXQ 1.0?
    2. If this is a muiltpart request with 1 part, what is the type of `$body`?
    3. If it is a multipart request with 1+n parts, what is the type of `$body`?
    4. As you have specified `cardinality MUST allow for zero or more`, have we lost the facility (and compatibility perhaps) from RESTXQ 1.0 for users to enforce through API design that this resource function only matches a request with a non-empty body?

2. or you can write this:
    ```xquery
    declare
       %reset:POST("{$body-part-1}", "{$body-part-2}", "{$body-part-3}")
    function my:example1($body-part-1 as item()*, $body-part-2 as item()*, $body-part-3 as item()*) {
    ...
    }
    ```

    1. If this is not a multipart request, what are the type of `$body-part-1`? Presumably `$body-part-2` and `$body-part-3` are then also empty sequences?
    2. If this is a muiltpart request with 1 part, what is the type of `$body-part-1`? Presumably `$body-part-2` and `$body-part-3` are then also empty sequences?
    3. If it is a multipart request with 3 parts, what are the types of `$body-part-1`, `$body-part-2`, and `$body-part-3`?
    4. If it is a multipart request with 5 parts, what are the types of `$body-part-1`, `$body-part-2`, and `$body-part-3`?
    5. As you have specified `cardinality MUST allow for zero or more`, are we missing the facility for users to exactly enforce through API design that this resource function only matches a request with a certain number of parts?
    6. As you have specified `The remaining body parts are injected to the function parameter corresponding to the final string literal as an array.`, what happens if there is only a function with one parameter (e.g. just `$body-part-1`) when there is a multi-part request with 1 part, or with 1+n parts?

Reply:
> request body parts SHOULD be processed in reverse in order to preserve the last bodypart having the highest priority

I think having two different ordering complicates things. I think it is better to stick with the order that appears in the HTTP request in all cases.

Reply:
> So do I understand correctly, you can write this:
> 
> ```xq
> declare
>    %reset:POST("{$body}")
> function my:example1($body as item()*) {
> ...
> }
> ```
> 
>  1. If this is not a multipart request, presumably the type of `$body` stays as it was in RESTXQ 1.0?

Yes

> 2. If this is a muiltpart request with 1 part, what is the type of `$body`?

The same as if it were a non-multipart request; since the parameter is not an array, the implementation would cast the multipart array to a sequence (in this case a singleton sequence), and continue to inject the body through the function parameter as normal.

> 3. If it is a multipart request with 1+n parts, what is the type of `$body`?

As above, the array would be cast as a sequence, each value of which would be converted according to its content type in the same way as any single request body would.

> 4. As you have specified `cardinality MUST allow for zero or more`, have we lost the facility (and compatibility perhaps) from RESTXQ 1.0 for users to enforce through API design that this resource function only matches a request with a non-empty body?

You are right, but perhaps we ought to look more closely at what the cardinality requirements need to be?  With multipart support, should we allow for API designers to require any cardinality they wish?

Reply:
> > request body parts SHOULD be processed in reverse in order to preserve the last bodypart having the highest priority
> 
> I think having two different ordering complicates things. I think it is better to stick with the order that appears in the HTTP request in all cases.

OK, although I think it is worth knowing that for `mulipart/alternative`, the highest preference is the last body part, the lowest the first.  Would we be best pointing this out to prospective users as a note referencing [RFC2046](https://datatracker.ietf.org/doc/html/rfc2046#section-5.1.4)?